core/src/test/java/org/apache/iceberg/TestTableMetadata.java

```diff
@@ -1752,6 +1755,54 @@ public void testUpgradeMetadataThroughTableProperty(int baseFormatVersion, int n
         .containsExactly(entry("key2", "val2"));
   }
 
+  @ParameterizedTest
+  @EnumSource(
+      value = FileFormat.class,
+      mode = EnumSource.Mode.EXCLUDE,
+      names = {"PARQUET", "PUFFIN", "METADATA"})
+  public void testCreateMetadataVersionWithIncompatibleFileFormat(FileFormat targetFormat) {
+    assertThatThrownBy(
+            () ->
+                TableMetadata.newTableMetadata(
+                    new Schema(),
+                    PartitionSpec.unpartitioned(),
+                    null,
+                    ImmutableMap.of(
+                        TableProperties.FORMAT_VERSION, String.valueOf(3),
+                        TableProperties.DEFAULT_FILE_FORMAT, targetFormat.name())))
+        .hasMessageContaining("file format is not supported in table format version");
```

Comment:
I would also add a `.isInstanceOf(IllegalArgumentException.class)` here and further below just to be clearer on what's being expected